src/misc/umap.c

```diff
@@ -0,0 +1,864 @@
+ *
+ * Note: Step 1 above involves deciding if two high-dimensional vectors "look similar"
+ *       which, because of the curse of dimensionality, is in many cases a highly
+ *       subjective and potentially controversial operation: thread with care and at
+ *       your own risk. Two high-dimensional vectors might look similar or extremely
+ *       different depending on the point of view/angle, and there are a lot of
+ *       viewpoints when the dimensionality ramps up.
+ *
+ * </para><para>
+ * References:
+ *
+ * </para><para>
+ * Leland McInnes, John Healy, and James Melville. https://arxiv.org/abs/1802.03426
+ *
+ * \param graph Pointer to the similarity graph to find a layout for (i.e. to embed).
+ * \param distances Pointer to a vector of edge lengths. Similarity graphs for
+ *   UMAP are often originally meant in terms of similarity weights (e.g. correlation between
+ *   high-dimensional vectors) and converted into distances by crude dist := 1 - corr. That is
+ *   fine here too. If this argument is a NULL pointer (NULL), all lengths are assumed equal.
+ * \param layout Pointer to the n x 2 matrix where the layout coordinates will be stored. Only 2D
+ *   embedding are currently supported, as they are by far more common than any higher dimensions.
+ * \param min_dist A fudge parameter that decides how close two unconnected vertices can be in the
+ *   embedding before feeling a repulsive force. It should be positive. Typically, 0.01 is a good
+ *   number.
+ * \param epochs Number of iterations of the main stochastic gradient descent loop on the
+ *   cross-entropy. Usually, 500 epochs can be used if the graph is the graph is small
+ *   (less than 50k edges), 50 epochs are used for larger graphs.
+ * \param sampling_prob The fraction of vertices moved at each iteration of the stochastic gradient
+ *   descent (epoch). At fixed number of epochs, a higher fraction makes the algorithm slower.
+ *   Vice versa, a too low number will converge very slowly, possibly too slowly.
+ *
+ * \return Error code.
+ *
+ */
+igraph_error_t igraph_layout_umap(const igraph_t *graph, const igraph_vector_t *distances,
+        igraph_matrix_t *layout, igraph_real_t min_dist, igraph_integer_t epochs, igraph_real_t sampling_prob) {
+
+
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    /* probabilities of each edge being a real connection */
+    igraph_vector_t umap_weights;
+    /* The smoothing parameters given min_dist */
+    igraph_real_t a, b;
+
+    if (min_dist <= 0) {
+        IGRAPH_ERRORF("Minimum distance should be positive, but found %f.",
+                IGRAPH_EINVAL, min_dist);
+    }
+
+    if (epochs < 0) {
+        IGRAPH_ERRORF("Number of epochs should be non-negative, but found %"
+                IGRAPH_PRId ".",
+                IGRAPH_EINVAL, epochs);
+    }
+
+    if (sampling_prob <= 0) {
+        IGRAPH_ERRORF("Sampling probability should be positive, but found %f.",
```

Comment:
Generally, use `%g` unless there is a specific reason for `%f`.

```suggestion
        IGRAPH_ERRORF("Sampling probability should be positive, but found %g.",
```

Reply:
yep, thanks